praw/util/refresh_token_manager.py

```diff
@@ -0,0 +1,79 @@
+"""Refresh Token Manager classes.
+
+There should be a 1-to-1 mapping betwen an instance of a subclass of
+:class:`~.BaseManager` and a :class:`~.Reddit` instance.
+
+A few trivial refresh token manager classes are provided here, but it is
+expected that PRAW users will create their own refresh token manager classes
+suitable for their needs.
+
+"""
+
+
+class BaseManager:
+    """An abstract class for all Refresh Token Managers."""
+
+    def __init__(self):
+        """Prepare attributes needed by all RefreshTokenManager classes."""
+        self._reddit = None
+
+    @property
+    def reddit(self):
+        """Return the :class:`~.Reddit` instance bound to the token manager."""
+        return self._reddit
+
+    @reddit.setter
+    def reddit(self, value):
+        if self._reddit is not None:
+            raise RuntimeError("reddit can only be set once and is done automatically")
+        self._reddit = value
+
+    def post_refresh_callback(self, authorizer):
+        """Handle callback that is invoked after a refresh token is used.
+
+        :param authorizer: The ``prawcore.Authorizer`` instance used containing
+        ``access_token`` and ``refresh_token`` attributes.
+
+        This function will be called after refreshing the access and refresh
+        tokens. This callback can be used for saving the updated
+        ``refresh_token``.
+
+        """
+        raise NotImplementedError("post_refresh_callback must be extended.")
+
+    def pre_refresh_callback(self, authorizer):
+        """Handle callback that is invoked before refreshing PRAW's authorization.
+
+        :param authorizer: The ``prawcore.Authorizer`` instance used containing
+        ``access_token`` and ``refresh_token`` attributes.
+
+        This callback can be used to inspect and modify the attributes of the
+        ``prawcore.Authorizer`` instance, such as setting the
+        ``refresh_token``.
```

Comment:
To ensure the stored refresh token hasn't already been used.

Reply:
Yes, that's a possibility, depending on the implementation. However, I'd consider always clearing out the refresh token in the `post_refresh_callback` after persisting it elsewhere so that `pre_refresh_callback` simply always fetches a valid one and doesn't need to do any comparison.